zerver/lib/queue.py

```diff
@@ -188,31 +189,32 @@ def _connect(self, on_open_cb: Optional[Callable[[], None]]=None) -> None:
         self.connection = ExceptionFreeTornadoConnection(
             self._get_parameters(),
             on_open_callback = self._on_open,
+            on_open_error_callback = self._on_connection_open_error,
             on_close_callback = self._on_connection_closed,
         )
 
     def _reconnect(self) -> None:
         self.connection = None
         self.channel = None
         self.queues = set()
+        self.log.warning("TornadoQueueClient attempting to reconnect to RabbitMQ")
         self._connect()
 
-    def _on_connection_closed(self, connection: pika.connection.Connection,
-                              reply_code: int, reply_text: str) -> None:
-        self.log.warning("TornadoQueueClient lost connection to RabbitMQ, reconnecting...")
-        from tornado import ioloop
-
-        # Try to reconnect in two seconds
-        retry_seconds = 2
+    CONNECTION_RETRY_SECS = 5
 
-        def on_timeout() -> None:
-            try:
-                self._reconnect()
-            except pika.exceptions.AMQPConnectionError:
-                self.log.critical("Failed to reconnect to RabbitMQ, retrying...")
-                ioloop.IOLoop.instance().call_later(retry_seconds, on_timeout)
+    def _on_connection_open_error(self, connection: pika.connection.Connection,
+                                  message: Optional[str]=None) -> None:
+        retry_secs = self.CONNECTION_RETRY_SECS
+        self.log.warning("TornadoQueueClient couldn't connect to RabbitMQ, retrying in %d secs..."
+                         % (retry_secs,))
```

Comment:
Why are you demoting from critical -> warning -- won't that impact whether we get error emails etc.?

Reply:
(It seems like if we fail twice in a row, which is what the old code required to do the `log.critical`, we probably do want to make noise about the issue)

Reply:
Hmm, I guess that happened because this bit of logic went away, and I copied the logging from `_on_connection_closed` instead. Sure, can make this critical.